File: typescript/systemdesign/pubSubEventBus.ts
/*
Simple publish/subscribe event bus supporting multiple topics, multiple
subscribers per topic, and unsubscribe handles - a lightweight in-process
message broker.

status - completed
*/

import { logp } from "../utils/logger";

type Handler<T> = (payload: T) => void;

class PubSubEventBus {
  private readonly subscribersByTopic: Map<string, Set<Handler<unknown>>> = new Map();

  subscribe<T>(topic: string, handler: Handler<T>): () => void {
    let subscribers = this.subscribersByTopic.get(topic);
    if (!subscribers) {
      subscribers = new Set();
      this.subscribersByTopic.set(topic, subscribers);
    }
    const wrapped = handler as Handler<unknown>;
    subscribers.add(wrapped);

    return () => {
      subscribers!.delete(wrapped);
    };
  }

  publish<T>(topic: string, payload: T): number {
    const subscribers = this.subscribersByTopic.get(topic);
    if (!subscribers || subscribers.size === 0) {
      return 0;
    }
    for (const handler of subscribers) {
      handler(payload);
    }
    return subscribers.size;
  }

  subscriberCount(topic: string): number {
    return this.subscribersByTopic.get(topic)?.size ?? 0;
  }
}

// usage scenarios

interface OrderEvent {
  orderId: number;
  status: string;
}

const bus = new PubSubEventBus();

const received: string[] = [];
const unsubscribeEmail = bus.subscribe<OrderEvent>("orders", (event) => {
  received.push(`email-service saw order ${event.orderId}: ${event.status}`);
});
bus.subscribe<OrderEvent>("orders", (event) => {
  received.push(`analytics-service saw order ${event.orderId}: ${event.status}`);
});
bus.subscribe<string>("logs", (line) => {
  received.push(`log-sink: ${line}`);
});

const deliveredCount = bus.publish<OrderEvent>("orders", { orderId: 42, status: "placed" });
logp(`published to "orders", delivered to ${deliveredCount} subscribers`);
received.forEach((line) => logp(line));
received.length = 0;

unsubscribeEmail();
const deliveredAfterUnsub = bus.publish<OrderEvent>("orders", { orderId: 42, status: "shipped" });
logp(`after unsubscribing email-service, delivered to ${deliveredAfterUnsub} subscriber(s)`);
received.forEach((line) => logp(line));

logp(`publishing to topic with no subscribers -> delivered to ${bus.publish("nobody-listening", "x")} subscribers`);

File: python/systemdesign/pub_sub_event_bus.py
"""
Publish/subscribe event bus supporting multiple topics, each with
multiple subscribers. Publishing a message on a topic fans it out to
every subscriber callback registered for that topic.

status - completed
"""

from collections import defaultdict

from base_logger.logging_event import create_logger


class PubSubEventBus:

    log = create_logger(__name__)

    def __init__(self):
        self.subscribers = defaultdict(list)
        self._next_subscription_id = 0
        self._subscription_index = {}

    def subscribe(self, topic, callback):
        subscription_id = self._next_subscription_id
        self._next_subscription_id += 1
        self.subscribers[topic].append((subscription_id, callback))
        self._subscription_index[subscription_id] = topic
        return subscription_id

    def unsubscribe(self, subscription_id):
        topic = self._subscription_index.pop(subscription_id, None)
        if topic is None:
            return False
        self.subscribers[topic] = [
            (sid, cb) for sid, cb in self.subscribers[topic] if sid != subscription_id
        ]
        return True

    def publish(self, topic, message):
        delivered = 0
        for _, callback in list(self.subscribers.get(topic, [])):
            callback(topic, message)
            delivered += 1
        return delivered


if __name__ == "__main__":
    log = create_logger("pub_sub_demo")

    bus = PubSubEventBus()
    received = {"orders": [], "payments": []}

    def order_handler_a(topic, message):
        received["orders"].append(("A", message))

    def order_handler_b(topic, message):
        received["orders"].append(("B", message))

    def payment_handler(topic, message):
        received["payments"].append(message)

    bus.subscribe("orders", order_handler_a)
    sub_b = bus.subscribe("orders", order_handler_b)
    bus.subscribe("payments", payment_handler)

    delivered = bus.publish("orders", "order-created:1001")
    log.info("published to 'orders', delivered to %d subscribers", delivered)
    log.info("received orders: %s", received["orders"])

    bus.publish("payments", "payment-captured:1001")
    log.info("received payments: %s", received["payments"])

    bus.unsubscribe(sub_b)
    received["orders"].clear()
    bus.publish("orders", "order-created:1002")
    log.info("after unsubscribing handler B, received orders: %s", received["orders"])
    log.info("expected only handler A's delivery")

    unrelated_delivered = bus.publish("unknown-topic", "noop")
    log.info("publish to topic with no subscribers -> delivered=%d", unrelated_delivered)
